app/uk/gov/hmrc/incometaxsessiondata/controllers/SessionController.scala

```diff
@@ -58,26 +55,75 @@ class SessionController @Inject() (
     // Here is required internalID => request.internalId and request.sessionId
     request.body.asJson
       .getOrElse(Json.obj())
-      .validate[SessionData] match {
-      case err: JsError               =>
+      .validate(Session.readsWithRequest(request)) match {
+      case err: JsError =>
         logger.error(s"[SessionController][set]: Json validation error while parsing request: $err")
         Future.successful(BadRequest(s"Json validation error while parsing request: $err"))
       case JsSuccess(validRequest, _) =>
-        save(validRequest)
+        handleValidRequest(validRequest)
           .recover { case ex =>
             logger.error(s"[SessionController][set]: Unexpected error while setting session: $ex")
             InternalServerError(s"Unexpected error while setting session: $ex")
           }
     }
   }
 
-  private def save(validRequest: SessionData) =
+  private def handleValidRequest(validRequest: Session): Future[Result] = {
```

Comment:
again, please move these private methods to service level, we should not expose this logic on the controller level;
we should call single service level method which will handle all these scenarios

Reply:
Oh. So the service would return Future[Result]? I assumed returning this in a service would be bad practice. I don't think I've seen it before

Reply:
as discussed last week, you can return Future[Either[Throwable,ResultOfOperation]], where is ResultOfOperation - would be type you have to design to model  all variation of outcomes for the given operation

Reply:
I meant in methods like handleValidRequest. All handleValidRequest/handleConflictScenario/handleOkScenario methods are doing is making calls to see if duplicate, then calling sessionService.set and handling that. I can move these to the service but they would have return type Future[Result]...

Reply:
Future[Result] is specific to the controller level, I don't think you need to return this type out of service, this would make testing much easier

Reply:
What would I return from the service then? All the controller methods were doing was calling the get and set methods then returning the results... They only looked _big_ because there was a lot of logging. Without logging it's like 15 lines in total

Reply:
as it been mentioned above, you can always return Future[Either[Throwable,ResultOfOperation]]